hazelcast/proxy/cp/fenced_lock.py

```diff
@@ -0,0 +1,533 @@
+
+        Returns:
+            hazelcast.future.Future[bool]: ``True`` if the lock was acquired and ``False``
+                if the waiting time elapsed before the lock was acquired.
+
+        Raises:
+            LockOwnershipLostError: If the underlying CP session was
+                closed before the client releases the lock
+        """
+        return self.try_lock_and_get_fence(timeout).continue_with(lambda f: f.result() != self.INVALID_FENCE)
+
+    def try_lock_and_get_fence(self, timeout=0):
+        """Acquires the lock if it is free within the given waiting time,
+        or already held by the current thread at the time of invocation and,
+        the acquire limit is not exceeded, and returns the fencing token
+        assigned to the current thread for this lock acquire.
+
+        If the lock is acquired reentrantly, the same fencing token is returned.
+        If the lock acquire limit is exceeded, then this method immediately returns
+        :const:`INVALID_FENCE` that represents a failed lock attempt.
+
+        If the lock is not available then the current thread becomes disabled
+        for thread scheduling purposes and lies dormant until the lock is
+        acquired by the current thread or the specified waiting time elapses.
+
+        If the specified waiting time elapses, then :const:`INVALID_FENCE`
+        is returned. If the time is less than or equal to zero, the method does
+        not wait at all. By default, timeout is set to zero.
+
+        This is a convenience method for the following pattern ::
+
+            lock = client.cp_subsystem.get_lock("lock").blocking()
+            if lock.try_lock():
+                fence = lock.get_fence()
+            else:
+                fence = lock.INVALID_FENCE
+
+        See Also:
+            :func:`lock_and_get_fence` function for more information about fences.
+
+        Args:
+            timeout (int): The maximum time to wait for the lock in seconds.
+
+        Returns:
+            hazelcast.future.Future[int]: The fencing token if the lock was acquired and
+                :const:`INVALID_FENCE` otherwise.
+
+        Raises:
+            LockOwnershipLostError: If the underlying CP session was
+                closed before the client releases the lock
+        """
+        current_thread_id = thread_id()
+        invocation_uuid = uuid.uuid4()
+        timeout = max(0, timeout)
+        return self._do_try_lock(current_thread_id, invocation_uuid, timeout)
+
+    def unlock(self):
+        """Releases the lock if the lock is currently held by the current thread.
+
+        Returns:
+            hazelcast.future.Future[None]:
+
+        Raises:
+            LockOwnershipLostError: If the underlying CP session was
+                closed before the client releases the lock
+            IllegalMonitorStateError: If the lock is not held by
+                the current thread
+        """
+        current_thread_id = thread_id()
```

Comment:
That is indeed a very good question. We can make the API blocking as you suggested in the above comment and that would simplify the solution, but it would be really weird if we have just the blocking version for FencedLock, and two versions for other proxies.

Therefore, I am proposing implementing a new type of future. It would be similar to the deferred version of the `boost::future` from the C++. We will lazily evaluate the callbacks on the thread that calls the `result()` method on the future objects and make the methods on this proxy to return this new future type.

So, when someone does this
```python
lock = client.cp_subsystem.get_lock("lock")
future = lock.lock()
future.add_done_callback(lambda _: print("callback"))
``` 

The print statement will not be executed even if the `future` resolves.

However, when someone does this
```python
lock = client.cp_subsystem.get_lock("lock")
future = lock.lock()
future.add_done_callback(lambda _: print("callback"))

...

future.result()
``` 
The callback would be executed in the thread that calls the `result` (main thread in most cases) after future resolves.

What do you think about this approach?

Reply:
What are the benefit in this approach when compared with the blocking API? The user will have to make sure to call `future.result()` on the same thread, which is more or less the same as calling a blocking `lock` method. Also, I wonder if this approach with lazy future execution will lead to user confusion and possible issues.

Reply:
Nothing comes to my mind apart from the consistent API offered by the client. 

It can in fact lead to confusion, and maybe we can implement the deferred future, provide a helper function that will convert normal futures to deferred ones, and return normal futures in the proxy?

So, if someone wants to unlock in a callback, he may do something like this

```python
from hazelcast.future import to_deferred

...

lock = client.cp_subsystem.get_lock("lock")
future = to_deferred(lock.lock())
future.add_done_callback(lambda _: print("callback"))

...

future.result()
```

Reply:
I'm not totally against this `to_deferred` approach and I understand the point of having a consistent API. My main concern is confusion that users familiar with `RLock` may have. If we decide to leave future-based API in `FencedLock`, we should clearly describe the underlying thread id logic, including some code snippets, in both ref manual and API docs.

Reply:
Sure, let me prepare the implementation and the documentation, and we can discuss it more with concrete examples

Reply:
@puzpuzpuz, after thinking about this problem over the weekend, I have decided to revert the changes done for the `deferred` idea, as it offers almost no value, besides having a somewhat consistent API.

But, I think it still makes sense to provide a non-blocking API. Maybe it can be useful for use-cases like below
```python

lock = ...
queue = ...

lock.lock().result()

def cb(_):
    queue.offer(3)

lock.unlock().add_done_callback(cb)
```

Anyways, I have added a note to the docstring mentioning that we advise using this proxy in blocking mode, non-blocking usage may cause a failure because the requests and callbacks may be executed on different threads.

Reply:
If there will be a warning in the ref manual and API doc, it should probably be fine to keep futures around.

> Anyways, I have added a note to the docstring mentioning that we advise using this proxy in blocking mode, non-blocking usage may cause a failure because the requests and callbacks may be executed on different threads.

Does it make sense to add some tests for non-blocking mode? Currently we only test the blocking one.

Reply:
Well, it is hard and unreliable to test those. For example, for `lock.unlock().add_done_callback(cb)`, the `cb` might be executed on the reactor thread if the `unlock` future is not completed, or on the thread that calls `add_done_callback`, if the future is already completed. We might make use of some internals such as `reactor.add_timer()` to execute everything on the reactor thread, but I would rather not.

We do have mock tests for all kinds of failure scenarios, I believe these should be enough

Reply:
> Well, it is hard and unreliable to test those.

If it's hard for us to test this code, how the users will feel?

Reply:
I meant testing it that way is hard and unreliable due to the problem I mentioned(It is specific to APIs that rely on thread ids, for others, it is straightforward to test). 

Using it is pretty simple